### PR TITLE
Non-parallel GitHub downloads, one progress bar per download

### LINKED ISSUE #1085: TRAP THE ERRORSSSSSS!!!!!!
It's a huge pain in the ass to be downloading/installing mods, then get an unhandled exception error when some web page that supposedly allows the plugin to download gives a 403 error! Also, I'm getting unhandled errors when faced with contradictory mods that have been selected for install. It needs to be given a tab selection method for choosing between the two.

### LINKED ISSUE #1817: Mods fail to download, fail to error - indefinite wait for interrupt.
CKAN 0.18.0
Windows 7 SP1 x64
On installing a list of mods, partway through downloading, the download stops indefinitely. No error displayed, until clicking 'Cancel', when the following errors were displayed (sequentially).

Failed to download "https://github.com/Crzyrndm/B9-PWings-Fork/releases/download/0.40.7/B9-PWings-Fork.v0.40.7.zip" - error: An exception occurred during a WebClient request.

Failed to download "https://github.com/jrossignol/ContractPack-AnomalySurveyor/releases/download/1.6.0/ContractPack-AnomalySurveyor_1.6.0.zip" - error: The remote server returned an error: (403) Forbidden.

Failed to download "https://github.com/DaMichel/KerbalFlightData/releases/download/R21/KerbalFlightDataR21.zip" - error: The remote server returned an error: (403) Forbidden.

Ideally, under this circumstance it would be preferable to display an error message, rather than hanging indefinitely. 

The list of mods was:

About to install...
- Advanced Jet Engine v2.7.2
- Asteroid Day 1.1.3
- B9 Aerospace Procedural Wings - Fork 1:0.40.7
- Community Tech Tree 1:2.4
- Contract Pack: Anomaly Surveyor 1.6.0
- Filter Extensions - Plugin 2.6.0.1
- Filter Extensions - Stock Filter Rework 2.6.0.1
- Hide Empty Tech Tree Nodes 0.4
- IndicatorLights 0.10
- Kerbal Construction Time 1.3.4
- Kerbal Flight Data R21
- Kerbal NRAP - Procedural Test Weights 1.5.1.1
- Kerbal Inventory System 1.2.12.0
- Near Future Construction 0.6.4(cached)
- Near Future Electrical 0.7.5(cached)
- Near Future Propulsion 0.7.3(cached)
- Near Future Solar 0.6.2(cached)
- Procedural Fairings v3.17
- 6 Seat Mk3 Cockpit 1.1
- Filter Extensions - Default Configuration 2.6.0.1
- Solver Engines plugin v2.3
- MagiCore 1.1.1
- Near Future Electrical Core 0.7.5(cached)
- Near Future Solar Core 0.6.2(cached)

### LINKED ISSUE #2210: CKAN's GitHub downloads are breaking the rules
# GitHub downloading needs a rewrite

(I debated whether to add this as a comment to #1817, but it seems like too much text and detail for that.)

## Problems

Currently if CKAN downloads many files from GitHub at the same time, they often fail with HTTP status code 403-Forbidden. #1817 contains an example, but these reports are common and I've definitely seen it happen myself several times.

## Background

The GitHub API uses 403 codes for throttling; you get 60 unauthenticated requests per hour, and any beyond that return a 403. I encountered this while working on [an unrelated project](https://hebarusan.github.io/releases-timeline/), and I had to use a GitHub token to allow 5000/hour, passed in the HTTP request headers:

```http
Authorization: token <OAuth token here>
```

Currently CKAN's downloads do **not** go through the GitHub API, so this does not necessarily indicate exactly what's going on with them. However, it establishes that 403-Forbidden is sometimes used for throttling, and it becomes more relevant later in discussion of the API.

[Sample API data for releases](https://api.github.com/repos/HebaruSan/Astrogator/releases/7538924), minus the author and uploader fields since they're long and not relevant to this issue:

```json
{
  "url": "https://api.github.com/repos/HebaruSan/Astrogator/releases/7538924",
  "assets_url": "https://api.github.com/repos/HebaruSan/Astrogator/releases/7538924/assets",
  "upload_url": "https://uploads.github.com/repos/HebaruSan/Astrogator/releases/7538924/assets{?name,label}",
  "html_url": "https://github.com/HebaruSan/Astrogator/releases/tag/v0.7.8",
  "id": 7538924,
  "tag_name": "v0.7.8",
  "target_commitish": "master",
  "name": "Frictionless toilet",
  "draft": false,
  "prerelease": false,
  "created_at": "2017-08-28T06:00:32Z",
  "published_at": "2017-08-28T06:09:07Z",
  "assets": [
    {
      "url": "https://api.github.com/repos/HebaruSan/Astrogator/releases/assets/4682631",
      "id": 4682631,
      "name": "Astrogator.zip",
      "label": null,
      "content_type": "application/zip",
      "state": "uploaded",
      "size": 114372,
      "download_count": 5243,
      "created_at": "2017-08-28T06:08:01Z",
      "updated_at": "2017-08-28T06:08:02Z",
      "browser_download_url": "https://github.com/HebaruSan/Astrogator/releases/download/v0.7.8/Astrogator.zip"
    }
  ],
  "tarball_url": "https://api.github.com/repos/HebaruSan/Astrogator/tarball/v0.7.8",
  "zipball_url": "https://api.github.com/repos/HebaruSan/Astrogator/zipball/v0.7.8",
  "body": "Fix glitches when the settings file is invalid"
}
```

The zip file that we want to download is associated with `assets[0]`, and there are two fields for it, `url` and `browser_download_url`. This becomes important later.

## Investigation summary

I used the "Contact GitHub" link to reach out to GitHub about how their download throttling works. Surprisingly, the person who replied understood exactly what I was talking about and how to fix it :+1:. It turns out that these problems happen because CKAN is not using GitHub as intended. From my conversation with the very helpful support person:

> If I understood your message correctly, it seems like you're programmatically downloading resources from github.com, is that right? If that's so, then you shouldn't be doing that. GitHub.com wasn't build for programmatic use like that, it was built for humans. For programmatic use, you should be using the API. The API has well defined rate limits and caching behavior you can rely on, while GitHub.com doesn't. That doesn't mean that github.com doesn't have any rate limits, it only means that you can be rate limited at any time and without warning.
> 
> So, we'd like to ask you to switch and use the API for downloading the data you need, and respect the defined rate limits (that's what a good citizen app should be doing, instead of hitting github.com):

("good citizen" was my phrasing in my original message, so don't take that as an unprovoked criticism of our civic virtues.)

> If I'm interpreting that code snippet correctly, you're using the browser_download_url link, which, as the name suggests, is intended to be used by human users via a browser.
> 
> For downloading release assets via the API, you should be using this endpoint:
> 
> https://developer.github.com/v3/repos/releases/#get-a-single-release-asset
> 
> Notice this note: "To download the asset's binary content, set the Accept header of the request to application/octet-stream. The API will either redirect the client to the location, or stream it directly if possible. API clients should handle both a 200 or 302 response."
> 
> That would be the "url" field of a particular asset (which are listed when you fetch a release e.g. via https://developer.github.com/v3/repos/releases/#get-a-single-release), but with the addition of the special Accept header.

Key points:

- The URL from the field we're using currently (`browser_download_url`) is for users and browsers only, not applications. It can be throttled, but there is no explicit policy or workaround.
- We should be using the GitHub API for downloads. Currently we use it in the Netkan code that finds new releases, but for downloads we effectively impersonate a browser.
- This can be done by requesting the `url` field instead of `browser_download_url` and setting a custom HTTP header:

  ```http
  Accept: application/octet-stream
  ```

  I tested this with wget, and setting the Accept header did indeed give me the download. Without this header, it returns a JSON object describing the asset.

### GitHub-specific downloading metadata & logic

When downloading from GitHub, we need to send the custom HTTP header. This cannot be accomplished simply by swapping out the bad URL for the good URL in the `download` metadata field.

Proposed new metadata field:

- `github_download` - The `assets[0].url` value from the API

Specific changes:

- The spec/schema would need to be updated to allow this field.
- Netkan would need to be updated to generate this field.
- CKAN would need to be updated to check for the presence of this field, which would then trigger an alternate download method that sets the custom header.

### UI to handle 403 statuses

If a GitHub download returns a 403 status, we should handle the exception and notify the user that their downloads are being throttled. We could direct them to the setting (see below) and [web page dealing with GitHub auth tokens](https://github.com/settings/tokens), and/or advise them to wait 60 minutes for their limit to reset. https://api.github.com/rate_limit can be used to get the exact limit and timing numbers.

### GitHub token handling

Users will be limited to 60 GitHub downloads per hour, because this is the limit of the GitHub API. 140+ mod installs are pretty commonly mentioned on the forums, and reinstalling everything from scratch is a common method for dealing with compatible upgrades, so some users would probably encounter this limit and not appreciate the 60-minute wait to be able to download more. The only way around this is to use a GitHub auth token, which boosts the limit to 5000/hour per token.

It would be nice to ship a single internal auth token for all of CKAN, since then users would have the 5000/hour limit by default without having to worry about any of the details. More responses from the GitHub contact person:

> Including a single pre-defined token with the app so that this token is used by all users of the app is possible. You could create a scopeless token here https://github.com/settings/tokens/new and include that. A scopeless token doesn't have any special permissions -- it can be used for read-only access of public data. So, it would be safe in that way. However, someone could easily take your token from the app, and then drain the API quota for the user who owns the token by making lots of unnecessary API requests. At that point, the app would stop working for everyone who uses the app.

Deliberate abuse like that is unlikely, but assuming 200 downloads per active user per hour, a 5000/hour limit across all CKAN users would support 25 active users in a given hour. The number of active users at a given time isn't known, but the latest CKAN release has over 60000 downloads, so it's probably more than 25. If we were able to determine the limit we needed per hour, we could divide it by 5000 and then generate that many tokens and pick one randomly per request, but that might not be in the spirit of the API's rules.

#### A setting

We could create a new settings field called GitHub Auth Token, where the user could fill in their own tokens to allow more downloads. This could be instead of or in addition to any built-in tokens we may or may not use, and it should support all the UIs.

#### Multi-pass approach

1. Try with no authentication at all. This would succeed for the first 60 requests per user per hour, probably the majority but not all.
2. For the remaining requests that fail, retry with a single hard coded auth token. As long as we only use this as a fallback, the 5000/hour limit would only apply to downloads in excess the 60/hour.

### Migration concerns

If Netkan was updated to use this new scheme tomorrow, current CKAN clients would break unless the old `download` field was still populated. So we should not remove support for the old metadata immediately; GitHub downloads should use both `download` and `github_download` until all clients are updated.

## Or just download serially

The [API docs](https://developer.github.com/v3/guides/best-practices-for-integrators/) say:

> Make requests for a single user or client ID serially. Do not make requests for a single user or client ID concurrently.

So even with a token, CKAN's parallel download method would still be in violation of the letter of the law.

As a halfway measure, we could try scaling back the parallelization of downloads.

1. Check whether a download URL contains "github.com"
2. If so, add it to a pool of downloads to be handled serially
3. Handle all other downloads normally
4. When a download finishes, if it contains "github.com", then start a new download from the pool

This might solve the problem without messing with all the API/token stuff. We would still technically be misusing GitHub, but users should no longer encounter failed downloads as frequently.

### LINKED ISSUE #2400: Weird SSL(?) error
**CKAN Version:** 1.24.0
**KSP Version:** 1.3.1
**Operating System:** Debian  9
**Have you made any manual changes to your GameData folder (i.e., not via CKAN)?**  nope

Problem
-------
Tried to install Realism Overhaul for KSP1.3.1. 

That's a lot of mods, some of them come in rather large files. CKAN downloaded most files over the course of about 10 minutes, then stalled. I cancelled the install, and started over from the CKAN main menu, doing the same selections as before. Most mods were cached, but the three still outstanding downloads were failed with an certificate error.

Re-importing certificates as suggested in https://github.com/KSP-CKAN/CKAN/wiki/SSL-certificate-errors did not help (ca-certificates-mono is already the newest version, cert-sync has nothing to do).

It strikes me as odd that everything works fine for a while, then throws SSL errors two minutes later. The certificate on the github website is valid until 17-May-2018 (only six weeks to go); I don't think it has been replaced just now. I'm suspecting some load-balancing or abuse prevention mechanism is at work, hence the question mark in the title.

FWIW, here's the full output of the second (failed) download attempt. I'd like to emphasisze once more that all all "cached" files were downloaded just a few minutes before:

```
About to install...

 * Realism Overhaul v12.0.0 (cached)
 * Deadly Reentry Continued v7.6.2 (cached)
 * HangerExtender 3.5.2 (cached)
 * KSP AVC 1.1.6.2 (cached)
 * MechJeb 2 2.7.0.0 (cached)
 * Procedural Fairings v4.0 (cached)
 * Procedural Parts v1.2.14 (cached)
 * PersistentRotation 1.8.4 (cached)
 * KSC Switcher 0.8 (cached)
 * Realism Overhaul Craft Files v12.0.0 (cached)
 * RemoteTech v1.8.9 (cached)
 * TAC Life Support (TACLS) v0.13.6.0 (cached)
 * TextureReplacerReplaced V0.5.4 (cached)
 * Toolbar 1.7.16.5 (cached)
 * Ven's Stock Part Revamp v1.9.6 (cached)
 * DMagic Orbital Science 1.3.11 (cached)
 * Kerbal Attachment System 0.6.3.0 (cached)
 * SCANsat v18.4 (cached)
 * FASA 1:v7.2.2 (cached)
 * SXTContinued 1:0.3.23.1 (cached)
 * Advanced Jet Engine v2.10.0 (cached)
 * Ferram Aerospace Research 3:0.15.9.1 (github.com, 751.9 KB)
 * Solver Engines plugin v3.5 (cached)
 * Kerbal Joint Reinforcement v3.3.3 (cached)
 * RealChute Parachute Systems v1.4.5 (github.com, 2.6 MB)
 * Real Fuels rf-v12.5.0 (cached)
 * Community Resource Pack 0.8.1.0 (cached)
 * RealHeat v4.5 (cached)
 * Real Plume 2:v11.0.0 (cached)
 * SmokeScreen - Extended FX Plugin 2.7.6.1 (cached)
 * Background Resources v0.13.6.0 (cached)
 * Firespitter Core v7.6.0 (github.com, 29.2 MB)
 * Firespitter Resources config v7.6.0 (github.com, 29.2 MB)
 * Retractable Lifting Surface Module 0.1.5 (cached)

Downloading "https://github.com/ferram4/Ferram-Aerospace-Research/releases/download/v0.15.9.1_Liepmann/FAR_0_15_9_1_Liepmann.zip"
Downloading "https://github.com/StupidChris/RealChute/releases/download/v1.4.5/RealChute_v1.4.5.zip"
Downloading "https://github.com/snjo/Firespitter/releases/download/v7.6.0/Firespitter_7.6.0.zip"
Oh no! Our download failed with a certificate error!

Consult this page for help:
	https://github.com/KSP-CKAN/CKAN/wiki/SSL-certificate-errors
Error!
```

### PR DESCRIPTION
**Pull request by HebaruSan**
## Problem

If you try to install lots of mods that you haven't previously downloaded, you'll likely get errors like this (possibly in a different format since we've changed them a bunch recently):

> Failed to download "https://github.com/jrossignol/ContractPack-AnomalySurveyor/releases/download/1.6.0/ContractPack-AnomalySurveyor_1.6.0.zip" - error: The remote server returned an error: (403) Forbidden.

## Cause

GitHub doesn't like it when you try to download multiple things from it at the same time, which CKAN aggressively does. GitHub sometimes responds to such attempts with 403-Forbidden statuses (the precise rules are not publicly known).

The GitHub _API_ works similarly, except it has [defined](https://developer.github.com/v3/#abuse-rate-limits) (and [queryable](https://api.github.com/rate_limit)) throttling rules and explicitly [says not to download from it in parallel](https://developer.github.com/v3/guides/best-practices-for-integrators/#dealing-with-abuse-rate-limits):

> Make requests for a single user or client ID serially. Do not make requests for a single user or client ID concurrently.

## Motivation

Since we're planning a pre-release anyway, why not go nuts?

## Changes

Now if multiple GitHub downloads are requested, only one of them will be started, and the rest will wait. Non-GitHub downloads will still be performed in parallel as before. When the current GitHub download finishes, one of the others will be started, and so on until completion.

Now each active download has its own progress bar in GUI, so the user can tell what's happening:

image

Fixes #1817.
Fixes #2210. (URLs haven't been migrated to API format, but that's probably not needed.)
Probably fixes #2400.
Fixes #1085. (The description and conversation went all over the place, so it may not be possible to even define what the issue is about, but the original complaint was about 403 statuses during downloads.)

### CODE REVIEW

**Comment by HebaruSan** [comment]
In latest commit, we create a 3-second timer to remove completed progress bars and add new ones, so the display won't fill up with full bars, but it also won't be constantly bouncing around.

**Comment by DasSkelett** [comment]
After the timer patche, I get this exception from time to time:
```
System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
  at System.Collections.ArrayList+ArrayListEnumeratorSimple.MoveNext () [0x00013] in <b2b5eee6b3734e9586bbc3ab32e40bc7>:0
  at System.Windows.Forms.XplatUIX11.NextTimeout (System.Collections.ArrayList timers, System.DateTime now) [0x0003a] in <8637a6a20d8848c3b0482a80613e13f8>:0
  at System.Windows.Forms.XplatUIX11.UpdateMessageQueue (System.Windows.Forms.XEventQueue queue, System.Boolean allowIdle) [0x0009f] in <8637a6a20d8848c3b0482a80613e13f8>:0
  at System.Windows.Forms.XplatUIX11.UpdateMessageQueue (System.Windows.Forms.XEventQueue queue) [0x00000] in <8637a6a20d8848c3b0482a80613e13f8>:0
  at System.Windows.Forms.XplatUIX11.GetMessage (System.Object queue_id, System.Windows.Forms.MSG& msg, System.IntPtr handle, System.Int32 wFilterMin, System.Int32 wFilterMax) [0x0001c] in <8637a6a20d8848c3b0482a80613e13f8>:0
  at System.Windows.Forms.XplatUI.GetMessage (System.Object queue_id, System.Windows.Forms.MSG& msg, System.IntPtr hWnd, System.Int32 wFilterMin, System.Int32 wFilterMax) [0x00000] in <8637a6a20d8848c3b0482a80613e13f8>:0
  at System.Windows.Forms.Application.RunLoop (System.Boolean Modal, System.Windows.Forms.ApplicationContext context) [0x0034e] in <8637a6a20d8848c3b0482a80613e13f8>:0
  at System.Windows.Forms.Application.Run (System.Windows.Forms.ApplicationContext context) [0x00011] in <8637a6a20d8848c3b0482a80613e13f8>:0
  at System.Windows.Forms.Application.Run (System.Windows.Forms.Form mainForm) [0x00006] in <8637a6a20d8848c3b0482a80613e13f8>:0
  at CKAN.Main..ctor (System.String[] cmdlineArgs, CKAN.KSPManager mgr, System.Boolean showConsole) [0x002bc] in /home/kilian/git/CKAN/GUI/Main/Main.cs:164
  at at (wrapper remoting-invoke-with-check) CKAN.Main..ctor(string[],CKAN.KSPManager,bool)
  at CKAN.GUI.Main_ (System.String[] args, CKAN.KSPManager manager, System.Boolean showConsole) [0x00047] in /git/CKAN/GUI/Program.cs:35
  at CKAN.GUI.Main (System.String[] args) [0x00001] in /git/CKAN/GUI/Program.cs:15
```

Unfortunately the stack trace doesn't really help locating the cause. But I suspect it's one of the two foreach loops. Maybe I can track it down further.

File: GUI/Controls/Wait.cs [hunk 1]
```diff
@@ -44,10 +58,137 @@ public bool ProgressIndeterminate
...
+                        Style    = ProgressBarStyle.Continuous,
+                    };
+                    moduleBars.Add(module, newPb);
+                }
+                progressTimer.Start();
+            });
+        }
+
+        /// <summary>
+        // Add new progress bars and remove completed ones (100%) in a single scheduled pass,
```
Comment (DasSkelett) [inline GUI/Controls/Wait.cs:109]: One `/` missing here. My IDE complained about a broken docstring :P

**Comment by HebaruSan** [comment]
Added a `ToList` to the first loop to maybe avoid that enumerator problem.

**Comment by DasSkelett** [comment]
Just wanted to ask if this is intentional or not, but I like it:
Progress bars for downloads that complete during the first three seconds are not shown since the timer also waits 3 seconds for the first trigger.

But I think that's quite nice, because you probably aren't interested in the progress bars of downloads which complete that fast, only for ones that take longer.

**Comment by HebaruSan** [comment]
> Progress bars for downloads that complete during the first three seconds are not shown since the timer also waits 3 seconds for the first trigger.

Partially intentional. :)

To show all the downloads in the first 3 seconds, we would have to restore the code that displays the progress bars on the fly, and the annoying flickering would be back (even if just for 3 seconds). And the first 3 seconds aren't quite special—if a download both starts _and_ ends in _any_ interval, it won't get a progress bar.

**Pull request merged**